labgrid/util/ssh.py

```diff
@@ -0,0 +1,324 @@
+        for name, connection in self._connections.items():
+            if connection.isconnected():
+                connection.disconnect()
+        names = self._connections.copy().keys()
+        for name in names:
+            self.remove_by_name(name)
+
+
+def check_connected(func):
+    """Check if an SSHConnection is connected as a decorator"""
+
+    @wraps(func)
+    def wrapper(cls, *_args, **_kwargs):
+        if not cls.isconnected():
+            raise ExecutionError(
+                "{} can not be called ({} is not connected)".format(
+                    func.__qualname__, cls
+                )
+            )
+        return func(cls, *_args, **_kwargs)
+
+    return wrapper
+
+
+@attr.s
+class SSHConnection:
+    """SSHConnections are individual connections to hosts managed by a control
+    socket. In addition to command execution this class also provides an
+    interface to manage port forwardings. These are used in the remote
+    infrastructure to tunnel multiple connections over one SSH link.
+
+    A public identity infrastructure is assumed, no extra username or passwords
+    are supported."""
+    host = attr.ib(validator=attr.validators.instance_of(str))
+    _connected = attr.ib(
+        default=False, validator=attr.validators.instance_of(bool)
+    )
+    _tmpdir = attr.ib(
+        default=attr.
+        Factory(lambda: tempfile.mkdtemp(prefix="labgrid-connection-")),
+        validator=attr.validators.instance_of(str)
+    )
+    _forwards = attr.ib(default=attr.Factory(dict))
+
+    def __attrs_post_init__(self):
+        self._logger = logging.getLogger("{}".format(self))
+        self._ssh_prefix = "-o LogLevel=ERROR"
+        self._ssh_prefix = "-o PasswordAuthentication=no"
+        self._socket = os.path.join(
+            self._tmpdir, 'control-{}'.format(self.host)
+        )
+
+    def _open_connection(self):
+        """Internal function which appends the control socket and checks if the
+        connection is already open"""
+        self._ssh_prefix += " -o ControlPath={}".format(
+            self._socket
+        ) if self._check_master() else ""
+
+    def _run_socket_command(self, command, forward=""):
+        "Internal function to send a command to the control socket"
+        complete_cmd = "ssh -x -o ControlPath={cpath} -O {command}{forward} {host}".format(
+            cpath=self._socket,
+            command=command,
+            forward=forward,
+            host=self.host
+        ).split(' ')
+        res = subprocess.check_call(
+            complete_cmd,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+            timeout=2
+        )
+
+        return res
+
+    def _run_command(self, command):
+        "Internal function to run a command over the SSH connection"
+        complete_cmd = "ssh -x -o ControlPath={cpath} {host} {command}".format(
```

Comment:
Build the command as a list to avoid problems with spaces in the arguments.

Reply:
same as above